File: LeetCode/Medium/0767-reorganize-string/0767-reorganize-string.cpp
class Solution {
public:
    string reorganizeString(string s) {
        unordered_map<char, int> freq;
        for(auto c:s){
            freq[c]++;
        }
        priority_queue<pair<int, char>> max_heap;
        for(auto elem: freq){
            max_heap.push({elem.second, elem.first});
        }

        if(max_heap.top().first > (s.size()+1)/2)
            return "";
        
        string result;
        int prev_cnt=0;
        char prev_char=' ';
        while(!max_heap.empty()){
            auto cnt_char = max_heap.top();
            max_heap.pop();
            result.push_back(cnt_char.second);
            
            if(prev_cnt > 0)
                max_heap.push({prev_cnt, prev_char});
            prev_cnt = cnt_char.first - 1;
            prev_char = cnt_char.second;
        }
        return result;
    }
};

File: LeetCode/Medium/0767-reorganize-string/0767-reorganize-string.py
class Solution:
    def reorganizeString(self, s: str) -> str:
        freq = {}
        for char in s:
            freq[char] = freq.get(char,0) + 1
        
        max_heap = []
        for char, cnt in freq.items():
            heapq.heappush(max_heap, (-cnt, char))
        
        if -max_heap[0][0] > (len(s)+1)/2:
            return ""

        ans = []
        prev_cnt, prev_char = 0, ''

        while max_heap:
            cnt, char = heapq.heappop(max_heap)
            ans.append(char)
            if prev_cnt < 0:
                heapq.heappush(max_heap, (prev_cnt, prev_char))
            
            prev_cnt, prev_char = cnt + 1, char
        return "".join(ans)